service/frontend/admin/handler_test.go

```diff
@@ -1619,3 +1621,950 @@ func TestGetReplicationMessages(t *testing.T) {
+			hcHandlerFunc: func(mock *history.MockClient) {
+			},
+			wantErr: true,
+		},
+		"normal request": {
+			input: &types.GetCrossClusterTasksRequest{
+				TargetCluster: "test-cluster",
+			},
+			hcHandlerFunc: func(mock *history.MockClient) {
+				mock.EXPECT().GetCrossClusterTasks(gomock.Any(), gomock.Any()).Return(nil, nil)
+			},
+			wantErr: false,
+		},
+		"return error": {
+			input: &types.GetCrossClusterTasksRequest{
+				TargetCluster: "test-cluster",
+			},
+			hcHandlerFunc: func(mock *history.MockClient) {
+				mock.EXPECT().GetCrossClusterTasks(gomock.Any(), gomock.Any()).Return(nil, assert.AnError)
+			},
+			wantErr: true,
+		},
+	}
+
+	for name, td := range tests {
+		t.Run(name, func(t *testing.T) {
+			goMock := gomock.NewController(t)
+			hcMock := history.NewMockClient(goMock)
+			if td.hcHandlerFunc != nil {
+				td.hcHandlerFunc(hcMock)
+			}
+			handler := adminHandlerImpl{
+				Resource: &resource.Test{
+					Logger:        testlogger.New(t),
+					MetricsClient: metrics.NewNoopMetricsClient(),
+					HistoryClient: hcMock,
+				},
+			}
+
+			_, err := handler.GetCrossClusterTasks(context.Background(), td.input)
+			if td.wantErr {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func Test_RespondCrossClusterTasksCompleted(t *testing.T) {
+	tests := map[string]struct {
+		input         *types.RespondCrossClusterTasksCompletedRequest
+		hcHandlerFunc func(mock *history.MockClient)
+		wantErr       bool
+	}{
+		"nil request": {
+			input:   nil,
+			wantErr: true,
+		},
+		"empty target cluster": {
+			input:   &types.RespondCrossClusterTasksCompletedRequest{},
+			wantErr: true,
+		},
+		"normal request": {
+			input: &types.RespondCrossClusterTasksCompletedRequest{
+				TargetCluster: "test-cluster",
+			},
+			hcHandlerFunc: func(mock *history.MockClient) {
+				mock.EXPECT().RespondCrossClusterTasksCompleted(gomock.Any(), gomock.Any()).Return(nil, nil)
+			},
+			wantErr: false,
+		},
+		"respond error": {
+			input: &types.RespondCrossClusterTasksCompletedRequest{
+				TargetCluster: "test-cluster",
+			},
+			hcHandlerFunc: func(mock *history.MockClient) {
+				mock.EXPECT().RespondCrossClusterTasksCompleted(gomock.Any(), gomock.Any()).Return(nil, assert.AnError)
+			},
+			wantErr: true,
+		},
+	}
+
+	for name, td := range tests {
+		t.Run(name, func(t *testing.T) {
+			goMock := gomock.NewController(t)
+			hcMock := history.NewMockClient(goMock)
+			if td.hcHandlerFunc != nil {
+				td.hcHandlerFunc(hcMock)
+			}
+			handler := adminHandlerImpl{
+				Resource: &resource.Test{
+					Logger:        testlogger.New(t),
+					MetricsClient: metrics.NewNoopMetricsClient(),
+					HistoryClient: hcMock,
+				},
+			}
+
+			_, err := handler.RespondCrossClusterTasksCompleted(context.Background(), td.input)
+			if td.wantErr {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func Test_SerializeRawHistoryToken(t *testing.T) {
```

Comment:
consider doing a round-trip test on serializations. It is usually easier too. If you can serialize a struct, then unserialise it and then assert that they're equal you'll catch a lot of bugs. Look at `common/testdatagen` if you want an easy way to make data for round-trip tests

Reply:
Right, it is much better than testing separately